File: tasker-web/main/src/app/common/model/Task.ts
import {User} from "./User";
import {Comment} from "./Comment";

export interface Task {
  id?: number,
  summary?: string,
  description?: string,
  sprintId?: number,
  state?: TaskState,
  position?: number,
  reporter?: User,
  assignee?: User,
  comments?: Comment[],
  type?: TaskType,
  estimation?: number
}

export enum TaskState {
  OPEN = 'OPEN',
  IN_PROGRESS = 'IN_PROGRESS',
  IN_REVIEW = 'IN_REVIEW',
  DONE = 'DONE'
}

export enum TaskType {
  DEFECT = "DEFECT",
  STORY = "STORY",
  TASK = "TASK",
}

File: tasker-app/src/main/java/com/hidorikun/tasker/model/entity/Task.java
package com.hidorikun.tasker.model.entity;

import com.hidorikun.tasker.model.enums.TaskState;
import com.hidorikun.tasker.model.enums.TaskType;
import lombok.Data;

import javax.persistence.*;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;

@Data
@Entity
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @NotBlank
    @Size(max = 255)
    private String summary;

    @Column(columnDefinition="TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    private Sprint sprint;

    @ManyToOne(fetch = FetchType.LAZY)
    private User reporter;

    @ManyToOne(fetch = FetchType.LAZY)
    private User assignee;

    @NotNull
    @Enumerated(EnumType.STRING)
    private TaskState state;

    private Long position;

    @NotNull
    @Enumerated(EnumType.STRING)
    private TaskType type;

    @Temporal(TemporalType.TIMESTAMP)
    private Date createdOn;

    @Temporal(TemporalType.TIMESTAMP)
    private Date closedOn;

    private Long estimation;

    @OneToMany(
            mappedBy = "task",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    private Set<Comment> comments;

    public Task() {
        this.state = TaskState.OPEN;
        this.comments = new HashSet<>();
    }

}
